### ISSUE TITLE
Propagate constant load/store offsets more efficiently

### ISSUE THREAD

**Issue by MaxGraey** [AssemblyScript/assemblyscript#32]
For example:
```ts
const index = 4;
store<i32>(ptr + <usize>index * sizeof<i32>(), value);
// emit as: (i32.store (i32.add (get_local $0) (i32.const 16)) (get_local $2))
```

Good optimize to this in compile time:
```ts
store<i32>(ptr, value, 4 * sizeof<i32>());
// emit as: (i32.store offset=16 (get_local $0) (get_local $2))
```

So may be this good doing inside load / store buildins.

That do you think?

**Comment by dcodeIO** [comment]
The compiler itself avoids doing optimizations that diverge from the source, and leaves these to Binaryen. Actually, I had a similar idea, but added the `constantOffset` parameter instead because of [this comment](https://github.com/WebAssembly/binaryen/issues/1364#issuecomment-358195481).

**Comment by MaxGraey** [comment]
As mentioned **kripken** "but you need to know more about the pointer $1 here". So it seems it hard to do on binaryen side but you have much more information about managed pointers.

**Comment by dcodeIO** [comment]
Still the compiler would have to know for sure that `ptr` doesn't overflow, which doesn't seem trivial to me considering that the value is passed in by the user. What the compiler already does is, when you access a field on a class, it uses its actual offset as the load/store's offset.

**Comment by MaxGraey** [comment]
`It should probably also do this for array accesses eventually`
It will be perfect! Actually this motivate open this proposal

**Comment by MaxGraey** [comment]
Closed via #1103
